File: Data Structures & Algorithms/search-for-word/submission-8.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int ROW = board.length;
        int COL = board[0].length;

        Set<String> visited = new HashSet<>();

        for(int r=0;r<ROW;r++) {
            for(int c=0;c<COL;c++) {
                if(dfs(board, word, r, c,  0, visited)) {
                    return true;
                }
            }
        }

        return false;
    }


    public boolean dfs(char[][] board, String word, int r, int c, int i, Set<String> visited) {
        if(i == word.length()) {
            return true;
        }

        if(r < 0 || c < 0 || r >= board.length || c >= board[0].length || board[r][c] != word.charAt(i) || visited.contains(r + "." + c)) {
             return false;
        }

        // make the set value visited 
        visited.add(r + "." + c);

        boolean res = dfs(board, word, r+1,c, i+1, visited) || dfs(board, word, r-1,c, i+1, visited) || dfs(board, word, r,c+1, i+1, visited) || dfs(board, word, r,c-1, i+1, visited);

        // make the cell unvisited

        visited.remove(r + "." + c);
        return res;
    }
}

File: Data Structures & Algorithms/count-number-of-islands/submission-8.py
class Solution:
    def numIslands(self, grid: list[list[str]]) -> int:

        if len(grid) == 0:
            return 0

        ROW = len(grid)
        COL = len(grid[0])
        visited = set()
        count = 0
        for r in range(ROW):
            for c in range(COL):
                count += self.dfs(r,c, visited, grid, ROW, COL)

        
        return count

    def dfs(self, r, c, visited, grid, ROW, COL):
        if r >= ROW or c >= COL or r < 0 or c < 0 or (r,c) in visited or grid[r][c] == '0':
            return 0

        visited.add((r,c))

        self.dfs(r+1, c, visited, grid, ROW, COL)
        self.dfs(r-1, c, visited, grid, ROW, COL)
        self.dfs(r, c+1, visited, grid, ROW, COL)
        self.dfs(r, c-1, visited, grid, ROW, COL)

        return 1

        





        